File: linked_list/mergeTwoLists.c
#include "lists.h"


/**
 * mergeTwoLists - merges two sorted ListNode linked lists
 * @list1: address of the head node of the first list
 * @list2: address of the head node of the second list
 *
 * Return: On success, the address of the new merged sorted list.
*/
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2)
{
    struct ListNode *mergedHead, *mergedTail, *temp;

    if (!list1 && !list2)
        return (NULL);

    if (!list1 && list2)
        mergedHead = list2, list2 = list2->next;
    else if (list1 && !list2)
        mergedHead = list1, list1 = list1->next;
    else if (list1->val <= list2->val)
        mergedHead = list1, list1 = list1->next;
    else
        mergedHead = list2, list2 = list2->next;

    mergedTail = mergedHead;
    while (list1 != NULL && list2 != NULL)
    {
        if (list1->val <= list2->val)
            temp = list1, list1 = list1->next;
        else
            temp = list2, list2 = list2->next;

        mergedTail->next = temp;
        mergedTail = temp;
    }

    if (list1)
        mergedTail->next = list1;
    else if (list2)
        mergedTail->next = list2;
    return (mergedHead);
}

File: linked_list/mergeTwoLists.py
#!/usr/bin/python3
"""LeetCode Problem #21 --> Merge Two Sorted Lists"""
from typing import List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: ListNode, list2: ListNode) -> ListNode:
        head, tail = None, None

        while list1 and list2:
            if list1.val >= list2.val:
                if head is None:
                    head = ListNode(list2.val)
                    tail = head
                else:
                    tail.next = ListNode(list2.val)
                    tail = tail.next
                    print('here')
                list2 = list2.next
            else:
                if head is None:
                    head = ListNode(list1.val)
                    tail = head
                else:
                    tail.next = ListNode(list1.val)
                    tail = tail.next
                list1 = list1.next

        while list1:
            if head is None:
                head = ListNode(list1.val)
                tail = head
            else:
                tail.next = ListNode(list1.val)
                tail = tail.next
            list1 = list1.next

        while list2:
            if head is None:
                head = ListNode(list2.val)
                tail = head
            else:
                tail.next = ListNode(list2.val)
                tail = tail.next
            list2 = list2.next
        return head
